src/org/traccar/database/DeviceManager.java

```diff
@@ -122,13 +138,16 @@ private void updateDeviceCache(boolean force) throws SQLException {
                     devicesById.remove(cachedDeviceId);
                 }
             }
-            for (String cachedDeviceUniqId : devicesByUniqueId.keySet()) {
-                if (!databaseDevicesUniqueIds.contains(cachedDeviceUniqId)) {
-                    devicesByUniqueId.remove(cachedDeviceUniqId);
+            for (String cachedDeviceUniqueId : devicesByUniqueId.keySet()) {
+                if (!databaseDevicesUniqueIds.contains(cachedDeviceUniqueId)) {
+                    devicesByUniqueId.remove(cachedDeviceUniqueId);
```

Comment:
Just notices another problem here.

I know it was here before, but I think if this condition happens, there would be an exception. You can't iterate through a collection and remove elements from it at the same time. Can you please double check. I think instead you should be using iterators which allow to remove elements.